nlprule/src/bin/run.rs

```diff
@@ -18,11 +20,12 @@ fn main() {
     env_logger::init();
     let opts = Opts::parse();
 
-    let tokenizer = Tokenizer::new(opts.tokenizer).unwrap();
-    let rules = Rules::new(opts.rules).unwrap();
+    let tokenizer = Arc::new(Tokenizer::new(opts.tokenizer).unwrap());
+    let mut rules = Rules::new(opts.rules, tokenizer.clone()).unwrap();
+    rules.spell_mut().options_mut().variant = Some(rules.spell().variant("en_GB").unwrap());
```

Comment:
I find this very very un-ergonomic API.

A few particular points to avoid:

* mutable options in public API
* structs in public lib API
* circular borrow fun (need `rules.spell_mut()` and `rules.spell()` at the same time)
* combine artifacts that don't seem necessarily connected (spellcheck rules vs grammer / tokenizer rules)

A more rustic API imho would be:

```rust
// explicit type annotations for clarity
let spellcheck: Spellcheck = Spellcheck::new(path_to_spellcheck_rules)?;
// alt:
let spellcheck: Spellcheck = Spellcheck::from_reader(fs::File::opn(path_to_spellcheck_rules)?)?;
spellcheck.add_hunspell_dict

let previous: Option<Spellcheck> = rules.set_spell(spellcheck);
// and
rules.modify_spellcheck(move |spellcheck| Ok(spellcheck.extend_whitelist(words_iter)))?
```

Was there a particular reason to not split the _spelling_ rules from the _tokenization_ rules? A third file would not hurt and avoid having to (`cargo-spellcheck` :tophat: ) issues with deployment size - 900kB extra is _a lot_ when the upper limit is 10MB.

Reply:
I'll think about this. In my opinion a mutable `SpellOptions` struct where all mutations are guaranteed to be valid and the spellchecker is updated when it is dropped is more erognomic than something like `rules.modify_spellcheck(move |spellcheck| Ok(spellcheck.extend_whitelist(words_iter)))?`. E.g. it seamlessly allows all `HashSet` operations on the whitelist, otherwhise that would have to be reimplemented as `*_whitelist` methods.

> mutable options in public API

Do you have some more information / source on this? I wouldn't have thought that this is bad.

> circular borrow fun (need rules.spell_mut() and rules.spell() at the same time)

this wouldn't compile, you need the speller to get a `Variant` (the `Variant ` is owned, it doesn't need the spellchecker), then you can use the `Variant ` to set the options. But I'm also not perfectly happy with that part.

> combine artifacts that don't seem necessarily connected (spellcheck rules vs grammer / tokenizer rules)

This is a very valid point. The reason is that `.correct()` and `.suggest()` on the `Rules` should easily be configurable to do both, grammar and spelling correction. The best fix would probably be some meta-struct to combine any amount of suggestion providers (where `Rules` and `Spellcheck` are each one suggestion provider) which is strongly related to #50. Defining and implementing that would take significantly more work than I want to do before nlprule can do spellchecking :slightly_smiling_face:

> A third file would not hurt

A third binary and a function `Rules.set_spell` really does seem like a good idea. But I would argue that it *decreases* ergonomics, at least as long as the `Spell` structure is still part of the `Rules`.

> when the upper limit is 10MB.

I'd just ask the crates.io team to increase the size so you don't have to worry about this (especially since (I think) you distribute some hunspell data as well). They do it on a case by case basis (https://github.com/rust-lang/crates.io/issues/195). I understand the concern, but I'd argue that it is already quite impressive to store multiple hundred thousand valid words + lemmas + part-of-speech tags, rules, a noun chunker etc. in < 10MB. An extra 900kb for spellchecking should very rarely be an issue.

Reply:
There's one alternative API moving some of the changes anticipated for #50 to this PR without going completely out of scope:

A new trait `Suggest`:

```rust
trait Suggest {
    fn suggest(&self, text: &str, tokenizer: &Tokenizer) -> Vec<Suggestion>; // or iterator
    fn correct(&self, text: &str, tokenizer: &Tokenizer) -> String {
        // default impl to just apply all the suggestions
    }
}
```

then `Rules` and a new separate struct `Spell` implement `Suggest`.  The spell struct implements `from_reader`, `new`, etc. like the `Rules` but with an additional argument to denote the variant: `Spell::new("path/to/bin", "en_GB")`.

There is one new struct `Correcter` with the fields:

```rust
struct Correcter {
   suggesters: Vec<Box<dyn Suggest>>
}
```

that can be constructed with

```rust
let correcter = Correcter::new(vec![
    Rules::from_reader(...).into(),
   Spell::from_reader(..., "en_GB").into()
])?; // implements `FromIterator` too
```

and implements `Suggest` as well. Getting the inner suggesters would then have to be done with `downcast-rs` and look like:

```rust
let rules = correcter.get::<Rules>(0)?;
// and mutably:
let spell = correcter.get_mut::<Spell>(1)?;
// I'm not 100% sure if this works with downcast-rs but I think it should
```

Extending the whitelist etc. would stay the same as before:

```rust
spell.options_mut().whitelist.push("nlprule".into());
```

But would not need the guard anymore (so no `Drop` impl - I agree that mutable options with some guard are problematic because they hide cost from the user, this way I really don't see an issue). Setting the variant is fallible and would thus not be part of the options:

```rust
spell.set_variant("en_US")?;
```

The nice thing about all of this is that nothing is a breaking change since the `Rules` keeps the same methods, there are just additional `Correcter` and `Spell` structs. There are multiple problems though:
- `.correct` and `.suggest` need the `Tokenizer` as second argument. These will by far be the most commonly used methods so I'd like them to be as simple as possible e.g. `fn correct(text: &str) -> String` and `fn suggest(text: &str) -> Vec<Sugggestion>` like in the current PR without having to explicitly keep the tokenizer around if you don't need it.
- The downcast magic happening in the `Correcter` is not ideal but I believe it can not be avoided.
- A third binary hurts ergonomics, especially because including a binary is quite verbose at the moment.

What do you think?